README.md
# site

[![Netlify Status](https://api.netlify.com/api/v1/badges/8e79b99b-f520-4f98-88a4-d59faff69c4c/deploy-status)](https://app.netlify.com/sites/tender-boyd-d507c4/deploys) ![License](https://img.shields.io/github/license/eventzimmer/site)

eventzimmer is an aggregation service that let's you display events aggregated from different channels. 

It is location-aware and the perfect fit for your community. If you'd like to have a centralized place to display everything, you've come to the right place.

The `site` repository contains all source code for the website displayed at [eventzimmer.de](https://eventzimmer.de)

## Getting started

We're always happy for suggestions, improvements, and hunting down bugs. You've made the first step by reading this `README`. Congratz :sparkles:

`site` is based on:
- [Vue](http://vuejs.org)
- [Bootstrap](https://getbootstrap.com/)
- [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)

Below instructions will give you a local version of `site` up and running, assuming that you have `node` and `npm` installed:

```
git clone https://github.com/eventzimmer/site.git
cd site/
npm install
VUE_APP_ENDPOINT=https://api.eventzimmer.de npm run serve
```

To build a production version of `site`:
```
VUE_APP_ENDPOINT=https://api.eventzimmer.de npm run build
cd dist/
python3 -m http.server # will serve a version of the website on http://localhost:8080
```

`site` is a `vue-cli` based project. A full reference can be found [in their documentation](https://cli.vuejs.org/config/).

If you feel ready for submitting your changes, please read our [contribution guidelines](https://github.com/eventzimmer/schema/blob/master/CONTRIBUTING.md).

All pull requests run tests through [Travis](https://travis-ci.org), and deploy to [Netlify](https://netlify.com). This ensures that both quality standards are met, and software also works in production.

On the upside you can test your pull request right from GitHub, without further ado.

### Running a local backend

You may have noticed the `VUE_APP_ENDPOINT` environment variable. This instructs `site` to load events from the given endpoint.
The default for this variable is `localhost:3000`. In order to use the local backend, please have a look at the [schema manual](https://github.com/eventzimmer/schema/blob/master/SETUP.md), in order to set it up.

### Running eslint

To check linting errors you have to execute the following command:
`npx standard`


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Make it possible to bookmark events
- Make it possible to change city based on geolocation
- Added anonymous events for read more and external links
- Added getinsights.io - privacy conform analytics

### Removed
- Removed any occurence of Matomo

## [0.2.0] - 2019-10-21 
### Added
- Added Twitter to footer
- Deactivate "no-console" command
- Added "node: true" to remove the linting errors
- Installed eslint dependencies
- Removed linting errors
- Added pull request template
- Added i18n for english
- Rearranged the data and event time
- Allowed 30 char values to appear
- Included an ```<hr>``` tage for better organization   


### Fixed
- Replaced visible-sm with d-md-none because it is the correct way in BS4
- Properly calculate categories even for locations without categories
- Only display the read-more button if there is actually more text

### Changed
- Replace Travis with GitHub actions

## [0.1.0] - 2019-08-01
### Added
- Add range selection
- Display categories for every event
- Make it possible to filter by category
- Add all available tags to all events based on locations
- Persist selected city
- Added a back to top button

### Changed
- Filter events based on range
- Use PostgREST to query for data
- Use Matomo with JavaScript and provide an explicit opt-out link in the privacy page.
- Exclude events that are older than PAST_HOUR_LIMIT hours from the current day
- Location badges do not overflow any longer
- Change read more button to be outlined / more visible
- Refactor events to be an individual component

### Removed
- Remove fixtures

## [0.0.3] - 2019-06-12
### Added
- Added privately hosted Matomo with disabled cookies
- Added font-awesome and instagram icon + link
- Added SEO meta tags
- Add badge with remaining time until event

### Changed
- Decode html containing description
- Only list events from the current day on
- Use badges for date and location name
- Remove unused bootstrap JS and use font awesome minified bundle
- Make all card titles links
- Change footer link color to be more accessible
- Add label for city select
- Add titles for impress and privacy

### Removed
- Removed Sentry because of privacy

## [0.0.2] - 2019-05-01
### Added
- Loading text
- Add Sentry error reporting

### Changed
- Skip waiting for workers

## [0.0.1] - 2019-04-27
### Added
- Make it possible to list events from API or fixtures
- Add card-based event view
- Display events of current and upcoming month
- Added impress and privacy as static html pages
- Added linting and tests and Travis
- Deployed via Netlify


src/store/index.js
import Vue from 'vue'
import Vuex from 'vuex'
import createPersistedState from 'vuex-persistedstate'
import { getMonth, addMonths, getDate } from 'date-fns'
import selection from './selection';

Vue.use(Vuex)

const PAST_HOUR_LIMIT = 3 // How many hours in the past would we like to include on the current day. In the future this may additionally depend on the length of the event.
const ENDPOINT = (process.env.VUE_APP_ENDPOINT !== undefined) ? process.env.VUE_APP_ENDPOINT : 'http://localhost:3000'

const LOCATIONS_ANNOTATED = require('@/assets/locations_annotated.json')
const CATEGORIES_PER_LOCATION = LOCATIONS_ANNOTATED
  .map((location) => {
    return {
      name: location.name,
      tags: location.tags.split(',').filter((part) => part.length)
    }
  })
  .reduce((accumulator, location) => {
    accumulator[location.name] = location.tags
    return accumulator
  }, {})

export default new Vuex.Store({
  state: {
    location: {
      name: 'Tübingen',
      latitude: 48.52,
      longitude: 9.05556
    },
    events: []
  },
  mutations: {
    changeEvents (state, events) {
      state.events = events
    },
    changeLocation (state, location) {
      state.location = location
    }
  },
  actions: {
    fetchEvents (context) {
      // eslint-disable-next-line no-undef
      return fetch(`${ENDPOINT}/rpc/events_by_radius?${new URLSearchParams({
        latitude: context.state.location.latitude,
        longitude: context.state.location.longitude,
        select: 'name,url,description,created_at,starts_at,location(*),source(*)',
        order: 'starts_at'
      })}`, {
        'Content-Type': 'application/json'
      }).then((response) => response.json())
        .then((events) => context.commit('changeEvents', events))
        .catch((err) => {
          throw err
        })
    }
  },
  getters: {
    nextMonthEvents (state, getters) {
      const nextMonth = getMonth(addMonths(new Date(), 1))
      return getters.eventsBySelectedCategories.filter((e) => getMonth(e.starts_at) === nextMonth)
    },
    currentMonthEvents (state, getters) {
      const today = new Date()
      const currentDay = getDate(today)
      const currentMonth = getMonth(today)
      return getters.eventsBySelectedCategories.filter((e) => (getMonth(e.starts_at) === currentMonth) && (getDate(e.starts_at) >= currentDay) && ((getDate(e.starts_at) === currentDay) ? (today.getHours() - e.starts_at.getHours() <= PAST_HOUR_LIMIT) : true))
    },
    eventsBySelectedCategories (state, getters) {
      if (state.selection.categories.length) {
        return getters.eventsWithCategories.filter((e) => e.categories.some((c) => state.selection.categories.includes(c)))
      } else {
        return getters.eventsWithCategories
      }
    },
    eventsByRange (state, getters) {
      return (state.selection.start !== null) ? getters.events.filter((e) => (e.starts_at >= getters.range.start) && (e.starts_at <= getters.range.end)) : getters.events
    },
    eventsWithCategories (state, getters) {
      return getters.eventsByRange.map((event) => {
        event.categories = CATEGORIES_PER_LOCATION[event.location.name]
        return event
      })
    },
    events (state) {
      return state.events.map((e) => {
        e.starts_at = new Date(e.starts_at)
        return e
      })
    },
    categories () {
      return new Set(Object.values(CATEGORIES_PER_LOCATION).reduce((acc, val) => acc.concat(val), [])) // NOTE: This can be replaced with Array.flat
    }
  },
  modules: {
    selection: selection
  },
  plugins: [
    createPersistedState({
      key: 'eventzimmer',
      paths: ['location', 'selection'],
      filter: (mutation) => !['changeEvents'].includes(mutation.type)
    })
  ]
})

src/store/selection.js
const selection = {
  state: {
    categories: [],
    bookmarks: [],
    cities: [],
    start: null,
    end: null,
    language: null
  },
  getters: {
    range (state) {
      return (state.start !== null) ? {
        start: new Date(state.start),
        end: new Date(state.end)
      } : null
    },
    bookmarks (state) {
      return new Set(state.bookmarks)
    }
  },
  mutations: {
    bookmarkEvent(state, url) {
      let bookmark = state.bookmarks.indexOf(url)
      if (bookmark !== -1) {
        state.bookmarks.splice(bookmark, 1)
      } else {
        state.bookmarks.push(url)
      }
    },
    addCity (state, city) {
      if (state.cities.findIndex((c) => c.name === city.name) === -1) {
        state.cities.push(city)
      }
    },
    changeCategories (state, categories) {
      state.categories = categories
    },
    changeRange (state, range) {
      if (range !== null) {
        state.start = range.start.toISOString()
        state.end = range.end.toISOString()
      } else {
        state.start = state.end = null
      }
    },
    changeLanguage (state, language) {
      state.language = language
    }
  },
}

export default selection;


src/store.js
import Vue from 'vue'
import Vuex from 'vuex'
import createPersistedState from 'vuex-persistedstate'

import { getMonth, addMonths, getDate } from 'date-fns'

Vue.use(Vuex)

const PAST_HOUR_LIMIT = 3 // How many hours in the past would we like to include on the current day. In the future this may additionally depend on the length of the event.
const ENDPOINT = (process.env.VUE_APP_ENDPOINT !== undefined) ? process.env.VUE_APP_ENDPOINT : 'http://localhost:3000'

const LOCATIONS_ANNOTATED = require('@/assets/locations_annotated.json')
const CATEGORIES_PER_LOCATION = LOCATIONS_ANNOTATED
  .map((location) => {
    return {
      name: location.name,
      tags: location.tags.split(',').filter((part) => part.length)
    }
  })
  .reduce((accumulator, location) => {
    accumulator[location.name] = location.tags
    return accumulator
  }, {})
  
const selection = {
  state: {
    categories: [],
    start: null,
    end: null,
    language: null
  },
  getters: {
    range (state) {
      return (state.start !== null) ? {
        start: new Date(state.start),
        end: new Date(state.end)
      } : null
    }
  },
  mutations: {
    changeCategories (state, categories) {
      state.categories = categories
    },
    changeRange (state, range) {
      if (range !== null) {
        state.start = range.start.toISOString()
        state.end = range.end.toISOString()
      } else {
        state.start = state.end = null
      }
    },
    changeLanguage (state, language) {
      state.language = language
    }
  }
}

export default new Vuex.Store({
  state: {
    location: {
      name: 'Tübingen',
      latitude: 48.52,
      longitude: 9.05556
    },
    events: []
  },
  mutations: {
    changeEvents (state, events) {
      state.events = events
    },
    changeLocation (state, location) {
      state.location = location
    }
  },
  actions: {
    fetchEvents (context) {
      // eslint-disable-next-line no-undef
      return fetch(`${ENDPOINT}/rpc/events_by_radius?${new URLSearchParams({
        latitude: context.state.location.latitude,
        longitude: context.state.location.longitude,
        select: 'name,url,description,created_at,starts_at,location(*),source(*)',
        order: 'starts_at'
      })}`, {
        'Content-Type': 'application/json'
      }).then((response) => response.json())
        .then((events) => context.commit('changeEvents', events))
        .catch((err) => {
          throw err
        })
    }
  },
  getters: {
    nextMonthEvents (state, getters) {
      const nextMonth = getMonth(addMonths(new Date(), 1))
      return getters.eventsBySelectedCategories.filter((e) => getMonth(e.starts_at) === nextMonth)
    },
    currentMonthEvents (state, getters) {
      const today = new Date()
      const currentDay = getDate(today)
      const currentMonth = getMonth(today)
      return getters.eventsBySelectedCategories.filter((e) => (getMonth(e.starts_at) === currentMonth) && (getDate(e.starts_at) >= currentDay) && ((getDate(e.starts_at) === currentDay) ? (today.getHours() - e.starts_at.getHours() <= PAST_HOUR_LIMIT) : true))
    },
    eventsBySelectedCategories (state, getters) {
      if (state.selection.categories.length) {
        return getters.eventsWithCategories.filter((e) => e.categories.some((c) => state.selection.categories.includes(c)))
      } else {
        return getters.eventsWithCategories
      }
    },
    eventsByRange (state, getters) {
      return (state.selection.start !== null) ? getters.events.filter((e) => (e.starts_at >= getters.range.start) && (e.starts_at <= getters.range.end)) : getters.events
    },
    eventsWithCategories (state, getters) {
      return getters.eventsByRange.map((event) => {
        event.categories = CATEGORIES_PER_LOCATION[event.location.name]
        return event
      })
    },
    events (state) {
      return state.events.map((e) => {
        e.starts_at = new Date(e.starts_at)
        return e
      })
    },
    categories () {
      return new Set(Object.values(CATEGORIES_PER_LOCATION).reduce((acc, val) => acc.concat(val), [])) // NOTE: This can be replaced with Array.flat
    }
  },
  modules: {
    selection: selection
  },
  plugins: [
    createPersistedState({
      key: 'eventzimmer',
      paths: ['location', 'selection'],
      filter: (mutation) => !['changeEvents'].includes(mutation.type)
    })
  ]
})
